### PR TITLE
Enable profiler schedule to early-exit on problems

### PR DESCRIPTION
**Pull request by scotts**
If Kineto, in the C++ layer, hit errors that caused the Kineto side to exit profiling early, the PyTorch side in Python never learned of this. In particular, the `torch.profiler.schedule` object would still step through its entire schedule, never having learned that Kineto tried to exit early. This is particularly bad if Kineto wanted to exit early during warmup, and the schedule still tries to go through more warmup and then a bunch of active phases.

This PR checks the newly exposed (https://github.com/pytorch/kineto/pull/1367) `isStopped()` API in Kineto during the `torch.profiler.schedule` state transitions. If it has been stopped, we enter a new schedule state: `DEVICE_STOPPED`.

Earlier implementations avoided adding a new state to the scheduling state machine, but they ran into the problem that the intended profiling behavior would no longer line up with the schedule the user specified. That is, we tried effectively fast-forwarding to `NONE` or `RECORD_AND_SAVE`. But if the user has a repeating schedule, we'd actually then end up with _more_ profiling.

Now, if `torch.autogra._is_kineto_stopped()` is true, we'll enter the state `DEVICE_STOPPED`. Transitioning _into_ `DEVICE_STOPPED` shuts down profiling. We stay in that state until the schedule would ordinarily go into either `NONE` or `WARMUP`. This behavior ensures:

* Users get a partial trace--which is particularly useful since there was enough of a problem during tracing to shut it down.
* Their specified schedule behaves as requested.
* We still appropriately shut down profiling when Kineto has told us cannot continue.

### CODE REVIEW

File: torch/profiler/profiler.py [hunk 1]
```diff
@@ -956,6 +956,24 @@ def step(self) -> None:
...
+                ProfilerAction.RECORD_AND_SAVE,
+            )
+            and self.current_action != ProfilerAction.NONE
+            and torch.autograd._is_stopped()
+        ):
+            warn(
+                "GPU activity collection was stopped early. "
+                f"Aborting profiling at step {self.step_num}."
+            )
+            self.current_action = ProfilerAction.NONE
```
Comment (ryanzhang22) [inline torch/profiler/profiler.py:975]: For the `ProfilerAction.RECORD` ->  `ProfilerAction.NONE` transition, it looks like we stop the trace but don't call `self._trace_ready` ([ref](https://github.com/scotts/pytorch/blob/84e3b609aa805ee3118487fe98262f3c5874df57/torch/profiler/profiler.py#L883C52-L883C56)). This seems like it would be a change in behavior as IIUC Kineto still processes the early-stopped trace and profiler would eventually step into `RECORD_AND_SAVE`.

There's also a couple of warning logs in the transition table from `WARMUP` -> `NONE` and `RECORD` -> `NONE` indicating that these are invalid (which could be confusing for reading the logs). Maybe we want to introduce a new transition state to handle these new cases instead?

File: torch/profiler/profiler.py [hunk 2]
```diff
@@ -956,6 +956,24 @@ def step(self) -> None:
         self.step_num += 1
         self.current_action = self.schedule(self.step_num)
 
+        # If GPU collection was stopped (e.g. buffer overflow), force an
+        # early stop regardless of what the schedule says.
+        if (
```
Comment (ryanzhang22) [inline torch/profiler/profiler.py:961]: Is there anything stopping the user from continuing to call `step()` in a loop and driving transitions? I'm imagining something like, GPU collection stops during warmup and we transition from `WARMUP` -> `NONE`, but then the user just calls `step` again and we transition right back from `NONE` -> `WARMUP` and re-do the instantiation logic. Because of the `prev_action` check here it doesn't seem like we'd be able to catch this case.

File: torch/profiler/profiler.py [hunk 3]
```diff
@@ -956,6 +956,24 @@ def step(self) -> None:
...
+            prev_action
+            in (
+                ProfilerAction.WARMUP,
+                ProfilerAction.RECORD,
+                ProfilerAction.RECORD_AND_SAVE,
+            )
+            and self.current_action != ProfilerAction.NONE
+            and torch.autograd._is_stopped()
```
Comment (ryanzhang22) [inline torch/profiler/profiler.py:969]: nit: imo `_is_kineto_stopped` or a similar name would make it more clear that this is a kineto-only concern. 

Related q: what happens if Kineto is available but the user only requested CPU activities?

File: torch/profiler/profiler.py (hunk 1 above)
Comment (scotts) [reply torch/profiler/profiler.py:975]: Good call out. I'll dig in.

File: torch/profiler/profiler.py (hunk 2 above)
Comment (scotts) [reply torch/profiler/profiler.py:961]: Nothing prevents that - I want to first figure out if we really support users directly calling `step()`. I'll dig in here as well.

File: torch/profiler/profiler.py (hunk 3 above)
Comment (scotts) [reply torch/profiler/profiler.py:969]: On `_is_kineto_stopped`: we're not consistent with using the `_kineto` prefix on calls down into it. We do it for `_kineto_step()`, but not for `_add_metadata_json()` and `_toggle_collection_dynamic()`. I think it's probably better to use `_is_kineto_stopped()`, but I want to explore the code more. We should probably be more consistent here; I don't think this layer is clear on what should be known where.

On only requesting CPU activities: I think you're right that the following sequence would be a problem in the current implementation:

1. User requests a trace with GPU. (CPU could be involved, but not needed.)
2. Kineto hits error, sets to stop the trace.
3. PyTorch exits schedule early.
4. User requests a new, CPU only trace.
5. Schedule would erroneously not work.

So I *think* we'll need to do some check on if we're collecting from GPUs. But I don't see us doing that kind of check yet, so I'll need to investigate more.

File: torch/profiler/profiler.py (hunk 2 above)
Comment (scotts) [reply torch/profiler/profiler.py]: Conclusion: it's not a problem. If the user (or the scheduler) calls `step()` again, we'll do the `NONE -> WARMUP` transition. That calls `prepare_trace()`, which calls into Kineto to enable tracing. That resets the flag we check in `_is_stopped()`.

File: torch/profiler/profiler.py (hunk 3 above)
Comment (scotts) [reply torch/profiler/profiler.py]: @ryanzhang22, agreed on the name, updated. I think the boundary between the PyTorch profiler and Kineto could probably use a refactor. At the least, I'm not sure what the abstraction layers are or even should be. I think that ideally we would have a layer in PyTorch that does not know "kineto" as a thing, it just knows there is a layer below that does device profiling. But I'm not sure where that layer should be. My current change, I think, is going in the wrong direction for that, but I also think it's the best thing to do until we can rethink how to abstract these layers.

File: torch/profiler/profiler.py [hunk 4]
```diff
@@ -963,6 +962,45 @@ def step(self) -> None:
...
+        #   profiling by a step when the schedule says to stop.
+        #
+        # When aborting from RECORD, we transition to RECORD_AND_SAVE rather
+        # than NONE so that the next step's RECORD_AND_SAVE -> NONE transition
+        # fires stop_trace + _trace_ready, saving the collected data. This
+        # costs one extra training step with profiling overhead.
+        if (
+            torch.autograd._is_kineto_stopped()
+            and self.use_device is not None
+            and prev_action
```
Comment (ryanzhang22) [inline torch/profiler/profiler.py]: If the user has code like:

```
with torch.profiler.profiler(
    schedule=(warmup_iters = 200)
) as prof:
    for iter in range(1000):
        self.model(...)
        prof.step()
```
(can't remember the exact param names)

and gpu collection stops for some reason on iter 100, but the user keeps calling `.step()`, is it valid to continue profiling on the step after that? I'm imagining something like:

1. Iter 100: _is_kineto_stopped = True, so we transition from `WARMUP` to `NONE`
2. Iter 101: we check .[schedule](https://github.com/scotts/pytorch/blob/9b2071b01ad5aace38c72882a600572f122e91e0/torch/profiler/profiler.py#L591C9-L591C13) and see that the state should be `WARMUP`, so we transition from `NONE` to `WARMUP`. In `_transit_action` this flips the collection stopped flag back to false and we continue stepping.

The semantics of "GPU collection stopped" makes it seem like we'd want to only allow profiling on a new profiling session, is it valid to allow Kineto to start back up in the same profiling loop?

Comment (scotts) [reply torch/profiler/profiler.py]: This is fine:

1. `WARMUP` -> `NONE` calls `start_trace()` and then `stop_trace()`.
2. `NONE` -> `WARMUP` calls `prepare_trace()`.

After `stop_trace()` we should be in the initial state as if we had done no profiling. This means that all of the state that caused the problem is reset. Yes, we may hit the same problem again (if there's something about the model that causes us to try to allocate too many buffers).

Because of how the scheduling works, transitioning through a full `WARMUP` -> `NONE` -> `WARUMP` cycle  _is_ effectively a new profiling session.

File: torch/profiler/profiler.py [hunk 5]
```diff
@@ -963,6 +1006,47 @@ def step(self) -> None:
...
+        # Entry into DEVICE_STOPPED:
+        # - use_device must be set: a CPU-only profiler must not react to a
+        #   stale stopped flag from a previous device profiler.
+        # - prev_action must not be NONE or DEVICE_STOPPED: if we already
+        #   transitioned out, the next cycle must be allowed to start.
+        # - current_action must not be NONE: if the schedule is already
+        #   stopping us (e.g. RECORD_AND_SAVE -> NONE), respect that rather
+        #   than entering DEVICE_STOPPED.
+        if prev_action == ProfilerAction.DEVICE_STOPPED:
+            if self.current_action not in (
```
Comment (ryanzhang22) [inline torch/profiler/profiler.py]: I think this could be a problem if warmup=0? If we have repeats it looks like the action would stay in DEVICE_STOPPED forever. 

We might want to release one step after the raw schedule reaches RECORD_AND_SAVE but that would require a third action variable to track the scheduled action before we overrode it in the last iteration.

Comment (scotts) [reply torch/profiler/profiler.py]: @ryanzhang22, yes, good call. We need to explicitly detect the cycle boundaries, which we now do now. And we do indeed another internal state variable to track what the scheduler actually told us (`self._prev_schedule_action`).

Added `warmup=0` and `wait=0` tests as well.

File: torch/profiler/profiler.py [hunk 6]
```diff
@@ -960,8 +1010,58 @@ def step(self) -> None:
...
+        # - use_device must be set: a CPU-only profiler must not react to a
+        #   stale stopped flag from a previous device profiler.
+        # - prev_action must not be NONE or DEVICE_STOPPED: if we already
+        #   transitioned out, the next cycle must be allowed to start.
+        # - current_action must not be NONE: if the schedule is already
+        #   stopping us (e.g. RECORD_AND_SAVE -> NONE), respect that rather
+        #   than entering DEVICE_STOPPED.
+        if prev_action == ProfilerAction.DEVICE_STOPPED:
+            at_cycle_boundary = (
+                self.current_action in (ProfilerAction.WARMUP, ProfilerAction.NONE)
```
Comment (ryanzhang22) [inline torch/profiler/profiler.py]: I think we want to remove `ProfilerAction.WARMUP` here, otherwise we'd get back into the 'WARMUP -> DEVICE_STOPPED -> WARMUP` loop in consecutive iterations (on the same profiling cycle). 

`prev_schedule_action == ProfilerAction.RECORD_AND_SAVE` is probably a good enough check, though `self.current_action == ProfilerAction.NONE` does help with weird schedules that don't use `ProfilerAction.RECORD_AND_SAVE` for some reason.

File: torch/profiler/profiler.py [hunk 7]
```diff
@@ -960,8 +1010,58 @@ def step(self) -> None:
...
+            if not at_cycle_boundary:
+                self.current_action = ProfilerAction.DEVICE_STOPPED
+        elif (
+            torch.autograd._is_kineto_stopped()
+            and self.use_device is not None
+            and prev_action
+            in (
+                ProfilerAction.WARMUP,
+                ProfilerAction.RECORD,
+                ProfilerAction.RECORD_AND_SAVE,
```
Comment (ryanzhang22) [inline torch/profiler/profiler.py]: When `RECORD_AND_SAVE` transitions to any other state (except None) we reset Kineto, so we can probably don't need to force `RECORD_AND_SAVE` to transition to `DEVICE_STOPPED`. Otherwise the cycle after would never be able to exit `DEVICE_STOPPED`.

File: torch/profiler/profiler.py [hunk 8]
```diff
@@ -941,6 +1017,20 @@ def __exit__(self, exc_type, exc_val, exc_tb):
             self.execution_trace_observer.cleanup()
 
     def start(self) -> None:
+        # Re-derive state from the schedule rather than trusting any
+        # carryover from a previous run. step() may have overridden
+        # current_action to DEVICE_STOPPED; if a user calls stop() then
+        # start() again, that stale DS would no-op the (NONE, DS)
+        # transition and leave the profiler running with no Kineto session.
+        schedule_action = self.schedule(self.step_num)
```
Comment (ryanzhang22) [inline torch/profiler/profiler.py]: nit: I think a more intuitive way to achieve this would be to reset `self.current_action` to `self._prev_schedule_action` in `stop()`. That way we don't need to call `schedule` twice.

**Review by ryanzhang22 — APPROVED** [review]
